lib/src/git_subprocess.rs

```diff
@@ -0,0 +1,759 @@
+        references: Vec<RefToPush>,
+    ) -> Result<(Vec<String>, Vec<String>), GitError> {
+        let args = {
+            let mut a = vec!["push".to_string(), "--porcelain".to_string()];
+
+            for reference in &references {
+                if let Some(refname) = reference.remote_reference() {
+                    a.push(format!(
+                        "--force-with-lease={refname}:{}",
+                        reference
+                            .expected_location
+                            .map(|x| x.to_string())
+                            .as_deref()
+                            .unwrap_or("")
+                    ));
+                }
+            }
+
+            a.extend_from_slice(&["--".to_string(), remote_name.to_string()]);
+
+            for reference in references {
+                a.push(reference.refspec.to_string());
+            }
+
+            a
+        };
+
+        let output = self.spawn(&args, Stdio::piped())?;
+
+        parse_git_push_output(output)
+    }
+}
+
+/// Generate a GitError::ExternalGitError if the stderr output was not
+/// recognizable
+fn external_git_error(stderr: &[u8]) -> GitError {
+    GitError::Subprocess(GitSubprocessError::External(format!(
+        "External git program failed:\n{}",
+        stderr.to_str_lossy()
+    )))
+}
+
+/// Parse no such remote errors output from git
+///
+/// To say this, git prints out a lot of things, but the first line is of the
+/// form:
+/// `fatal: '<branch_name>' does not appear to be a git repository`
+/// or
+/// `fatal: '<branch_name>': Could not resolve host: invalid-remote
+fn parse_no_such_remote(stderr: &[u8]) -> Result<(), String> {
+    if let Some(first_line) = stderr.lines().next() {
+        if (first_line.starts_with_str("fatal: '")
+            && first_line.ends_with_str("' does not appear to be a git repository"))
+            || (first_line.starts_with_str("fatal: unable to access '")
+                && first_line.ends_with_str("': Could not resolve host: invalid-remote"))
+        {
+            let mut split = first_line.split_str("\'");
+            split.next(); // ignore prefix
+            let branch_name = split.next();
+            if let Some(bname) = branch_name {
+                return Err(bname.to_str_lossy().to_string());
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Parse error from refspec not present on the remote
+///
+/// This returns
+///     Some(local_ref) that wasn't found by the remote
+///     None if this wasn't the error
+///
+/// On git fetch even though --prune is specified, if a particular
+/// refspec is asked for but not present in the remote, git will error out.
+///
+/// Git only reports one of these errors at a time, so we only look at the first
+/// line
+///
+/// The first line is of the form:
+/// `fatal: couldn't find remote ref refs/heads/<ref>`
+fn parse_no_remote_ref(stderr: &[u8]) -> Option<String> {
+    if let Some(first_line) = stderr.lines().next() {
```

Comment:
nit: `let first_line = stderr.lines().next()?`, etc. instead of `if let Some(..)`?